### DISCUSSION TITLE
Wormhole uses QUIC draft-29 which is deprecated in libp2p v0.30.0

### DISCUSSION THREAD

**Discussion by tbjump** [wormhole-foundation/wormhole#3348] [General] [answered]
Wormhole needs to upgrade to libp2p v0.30.0 in order to get golang 1.21 support. 
QUIC draft-29 is deprecated in libp2p v0.30.0, see https://github.com/libp2p/go-libp2p/releases/tag/v0.30.0
The switch to QUIC v1 is implemented in https://github.com/wormhole-foundation/wormhole/pull/3278 but this change would be incompatible with prior versions of the Guardian. This means that a quorum of Guardians and all subscribers would need to upgrade at the same time, which isn't great. Here are some migration strategy ideas instead: 

# Migration strategies
## (1) Dual-Stack transition period
* Run both QUIC draft-29 and QUIC v1 for some time. QUIC v1 would need to run on a new port. 
* Advise all clients to switch to QUIC v1, which would require updating their software and configuration
* Advise all Guardians to switch to a QUIC v1 bootstrap node, which would require updating their configuration
* Wait some time
* Remove support for QUIC draft-29 & upgrade to libp2p v0.30.0

Con: Requires software upgrade and configuration changes for guardians and subscribers; Requires exposing an additional port for Guardians

## (2) Timed concurrent upgrade
* Pick a time `t1` in the future
* Add code to the Guardian such that they automatically switch to QUIC v1 at exactly this point in time. 
* Advise all Guardians and subscriber to upgrade their software before `t1`. 
* Hope that clock-skew is not too large

## (3) ... other ideas? 
...

**Comment by tbjump** [comment]
After some offline discussion, it sounds like the Wormhole project is pursuing option (2) for now, since it requires less coordination between network participants. The next steps will be to develop a release candidate and test it thoroughly. 
There is an open question if the entire Guardian process needs to be restarted or if libp2p can be re-started cleanly.
If libp2p can be re-started cleanly, that would be the preferred option, to minimize network downtime, but libp2p used to have some issues with restarts in the past.

**Comment by bruce-riley** [comment, marked as answer]
Did some prototyping in both tilt and with `go test`.

Had to change references to `/quic` to `/quic-v1` and change the port number from `8999` to something else. I used `7999`.

- The newer version of QUIC seems to work with the current version of libp2p.
- The newer version of libp2p (v0.30.0) also seems to work with the above changes.

**Comment by SEJeff** [comment]
This has been deployed to testnet and will be deployed to mainnet January 16, 2024 thanks to the hard work of @bruce-riley.

**Comment by bruce-riley** [comment]
Mainnet cut over is set for January 16th, 2024.

**Reply by SEJeff** [reply]
Updated. Thanks

**Comment by bruce-riley** [comment]
Mainnet went live with `/quic-v1` on January 16, 2024. The next step is to upgrade the version of p2p used by the guardian. Once that is stable, we can upgrade to golang 1.21.

**Comment by aadam-10** [comment]
The cutover happened on Jan 16th as planned, closing discussion
